providers/implementations/signature/eddsa_sig.c

```diff
@@ -435,17 +630,94 @@ static int ed448_digest_verify(void *vpeddsactx, const unsigned char *sig,
 #endif /* S390X_EC_ASM */
 
     if (peddsactx->prehash_flag) {
-        if (!ed448_shake256(peddsactx->libctx, NULL, tbs, tbslen, md, mdlen))
+        if (!peddsactx->prehash_by_caller_flag) {
+            if (!ed448_shake256(peddsactx->libctx, NULL, tbs, tbslen, md, mdlen))
+                return 0;
+            tbs = md;
+            tbslen = mdlen;
+        } else if (tbslen != EDDSA_PREHASH_OUTPUT_LEN) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DIGEST_LENGTH);
             return 0;
-        tbs = md;
-        tbslen = mdlen;
+        }
+    } else if (peddsactx->prehash_by_caller_flag) {
+        /* The caller is supposed to set up a ph instance! */
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_INSTANCE);
+        return 0;
     }
 
     return ossl_ed448_verify(peddsactx->libctx, tbs, tbslen, sig, edkey->pubkey,
                              peddsactx->context_string, peddsactx->context_string_len,
                              peddsactx->prehash_flag, edkey->propq);
 }
 
+/* All digest_{sign,verify} are simple wrappers around the functions above */
+
+static int ed25519_digest_signverify_init(void *vpeddsactx, const char *mdname,
+                                          void *vedkey,
+                                          const OSSL_PARAM params[])
+{
+    PROV_EDDSA_CTX *peddsactx = (PROV_EDDSA_CTX *)vpeddsactx;
+
+    if (mdname != NULL && mdname[0] != '\0') {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DIGEST);
```

Comment:
I don't think this is the right error to report, it isn't that the digest is invalid per se, rather `EdDSA` (non-ph) does not support specifying a digest, so this probably deserves a more specific error.